src/app/server/JointFabricDatastore.cpp

```diff
@@ -127,185 +128,565 @@ CHIP_ERROR JointFabricDatastore::RemoveNode(NodeId nodeId)
+                                    Clusters::JointFabricDatastore::DatastoreStateEnum::kCommitted;
+                            }
+                        }));
+                        ++nkIt;
+                    }
+                }
                 else
                 {
-                    // TODO: update binding on actual device.
-                    // TODO: check if change has been made, retry if not. Mark as committed upon success.
-                    // TODO: retry delete, remove entry if successful.
+                    ++nkIt;
                 }
             }
         }
 
-        ++it;
-    }
+        // Request ACL List from the device and transition to kRefreshingACLs.
+        ReturnErrorOnFailure(mDelegate->FetchACLList(
+            mRefreshingNodeId,
+            [this](CHIP_ERROR err,
+                   const std::vector<Clusters::JointFabricDatastore::Structs::DatastoreACLEntryStruct::Type> & acls) {
+                if (err == CHIP_NO_ERROR)
+                {
+                    // Store fetched ACLs for processing in the next state.
+                    // TODO
+                    // mACLList = acls;
 
-    // TODO: read the Group Key Set List from the actual device
+                    // Advance the state machine to process the ACLs.
+                    mRefreshState = kRefreshingACLs;
+                }
+                else
+                {
+                    // Mark node as failed to refresh and tear down the refresh state.
+                    size_t idx = 0;
+                    if (IsNodeIDInDatastore(mRefreshingNodeId, idx) == CHIP_NO_ERROR)
+                    {
+                        mNodeInformationEntries[idx].commissioningStatusEntry.state =
+                            Clusters::JointFabricDatastore::DatastoreStateEnum::kCommitFailed;
+                        mNodeInformationEntries[idx].commissioningStatusEntry.failureCode = static_cast<uint8_t>(err.AsInteger());
+                    }
+                    mRefreshingNodeId = kUndefinedNodeId;
+                    mRefreshState     = kIdle;
+                }
 
-    // 4.
-    for (auto it = mGroupKeySetList.begin(); it != mGroupKeySetList.end();)
-    {
-        // TODO: Apply any pending updates from datastore, upon success mark as committed
-        // Retry any CommitFailure entries from datastore, upon success can be removed.
-        // Ensure list from device matches list in datastore
-        ++it;
+                // Continue the state machine to let the kRefreshingACLs branch process mACLList.
+                (void) ContinueRefresh();
+            }));
     }
-
-    // TODO: read the ACL List from the actual device
-
-    // 5.
-    for (auto it = mACLEntries.begin(); it != mACLEntries.end();)
-    {
-        if (it->nodeID == nodeId)
+    break;
+    case kRefreshingACLs: {
+        // 5.
+        for (auto it = mACLEntries.begin(); it != mACLEntries.end();)
```

Comment:
Similar to bindings, there is a single ACL attribute on this cluster with read/write access. As a result, what needs to happen here is:

1. Read the current list into mRefreshingACLList
2. Iterate through the mACLEntries for the given NodeId
3. If an entry is pending, then merge it into mRefreshingACLList
4. If an entry is deletePending, then remove it from mRefreshingACLList
5. Write mRefreshingACLList back to device
6. Replace mACLEntries for the given NodeId with entries from mRefreshingACLList

So in other words, there should only be one SyncNode call in the kRefreshingACLs case (at the end)